# Loops (Iteration Structure)

### What are loops?
- A loop is an iterative structure
- It helps to execute a block (code) repeatedly
- There are two types of loops:
    - Collection (countable) loops - `for` loop
    - Conditional loops - `while` loop

### The `for` loop
- This is a collection loop
- For every item in the collection the block (code) executes once, and after that loop ends
- Syntax:

```
for <var> in <collection>:
    block
```

In [28]:
for x in [10, 18, -25]:
    print("Hello, World!")

Hello, World!
Hello, World!
Hello, World!


In [29]:
for i in [10, -100]:
    print("Hello, World!")
    print("Good evening!")

Hello, World!
Good evening!
Hello, World!
Good evening!


In [30]:
for i in ["Naveen", "Kumar", "Nagesh"]:
    print("Hello, " + i + "!")
    print("---------------")

Hello, Naveen!
---------------
Hello, Kumar!
---------------
Hello, Nagesh!
---------------


In [31]:
for x in [10, 18, 25, 44, 19]:
    print("Temp:", x)

Temp: 10
Temp: 18
Temp: 25
Temp: 44
Temp: 19


In [32]:
for p in (100, -200, 300, -400, -500, 600):
    print("Value is:", p)
    print("--------------")
    print("Looping is fun!")

Value is: 100
--------------
Looping is fun!
Value is: -200
--------------
Looping is fun!
Value is: 300
--------------
Looping is fun!
Value is: -400
--------------
Looping is fun!
Value is: -500
--------------
Looping is fun!
Value is: 600
--------------
Looping is fun!


In [33]:
for x in [1001, "Raju", 35, "Hyd", "HR", 48000.0, True]:
    print("Item:", x)

Item: 1001
Item: Raju
Item: 35
Item: Hyd
Item: HR
Item: 48000.0
Item: True


In [34]:
for l in "Help":
    print("Char:", l)

Char: H
Char: e
Char: l
Char: p


In [35]:
for i in ("Naveen", 35, "Kumar", 40):
    print("Name or Age:", i)

Name or Age: Naveen
Name or Age: 35
Name or Age: Kumar
Name or Age: 40


In [36]:
total = 0
for t in [10, 15, 20, 30, 35]:
    total = total + t
print("Total is:", total)

Total is: 110


In [37]:
total = 1
for i in [5, 2, 7, 4, 3, 9]:
	total = total * i
print("Final Total is:", total)

Final Total is: 7560


In [38]:
for i in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]:
    print("18 x", i, "=", 18 * i)

18 x 1 = 18
18 x 2 = 36
18 x 3 = 54
18 x 4 = 72
18 x 5 = 90
18 x 6 = 108
18 x 7 = 126
18 x 8 = 144
18 x 9 = 162
18 x 10 = 180


In [39]:
fact = 1
for i in [1, 2, 3, 4, 5]:
    fact = fact * i
print("Factorial is:", fact)

Factorial is: 120


In [40]:
for i in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]:
    print("18 x", i, "=", 18 * i)

18 x 1 = 18
18 x 2 = 36
18 x 3 = 54
18 x 4 = 72
18 x 5 = 90
18 x 6 = 108
18 x 7 = 126
18 x 8 = 144
18 x 9 = 162
18 x 10 = 180
18 x 11 = 198
18 x 12 = 216
18 x 13 = 234
18 x 14 = 252
18 x 15 = 270
18 x 16 = 288
18 x 17 = 306
18 x 18 = 324
18 x 19 = 342
18 x 20 = 360


#### a list of series
1. 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
2. 0, 1, 2, 3, 4, ...
3. -10, -9, -8, -7, ..., 0, 1, 2, 3, ..., 10
4. 4, 3, 2, 1, 0, -1, -2, -3, -4
5. 7, 10, 13, 16, 19, 22, 25, ...
6. 4, 1, -2, -5, -8, -11, ...
7. 8, 18, 28, 38, 48, 58, ..., 108
8. 5 x 2 ** 1, 5 x 2 ** 2, 5 x 2 ** 3, 5 x 2 ** 4, 5 x 2 ** 5, ....
9. 1/3, 1/5, 1/7, 1/9

### The `range( [start,] stop [, step] )` function/iterator
- The `range()` iterator has 3 paramters: start, stop, step
- One required parameter: stop
- Two optional parameters: start and step
- The default value of start is 0
- The default value of step is 1
- The *range()* iterator returns values from 0 (start value) till stop value, by continuously adding step value (1). The stop value is exclusive.

#### The `range(stop)`
- Now start is 0, and step is 1
- Examples:
    * range(5)      # 0, 1, 2, 3, 4
    * range(3)      # 0, 1, 2
    * range(8)      # 0, 1, 2, 3, 4, 5, 6, 7

In [41]:
for i in range(5):
    print(i, end=" ")

0 1 2 3 4 

In [42]:
for i in range(3):
    print(i, end=" ")

0 1 2 

In [43]:
for i in range(8):
    print(i, end=" ")

0 1 2 3 4 5 6 7 

#### The `range(start, stop)` iterator
- By default step is 1; start < stop
- It will return one value for each iteration from start value to stop value, stop is exclusive, by adding step value

* Example:
    - range(10, 15)         # 10, 11, 12, 13, 14
    - range(-3, 3)          # -3, -2, -1, 0, 1, 2
    - range(1, 9)           # 1, 2, 3, 4, 5, 6, 7, 8
    - range(3, -3)          # wrong (not error)

In [44]:
for i in range(10, 15):
    print(i, end=" ")

10 11 12 13 14 

In [45]:
for i in range(-3, 3):
    print(i, end= " ")

-3 -2 -1 0 1 2 

In [46]:
for i in range(1, 9):
    print(i, end=" ")

1 2 3 4 5 6 7 8 

In [47]:
for i in range(3, -3):
    print(i, end=" ")

#### The `range(start, stop, step)` iterator
- start < stop; step is +ve
- start > stop; step is -ve

* Examples:
    - range(2, 20, 3)       # 2, 5, 8, 11, 14, 17
    - range(-20, 20, 8)     # -20, -12, -4, 4, 12
    - range(100, -30, -25)  # 100, 75, 50, 25, 0, -25
    - range(10, 50, -10)    # wrong (doesn't give an error, but doesn't work)
    - range(100, -100, -50) # 100, 50, 0, -50
    - range(100, -100, 50)  # wrong (doesn't give an error, not works)

, 

In [48]:
for i in range(2, 20, 3):
    print(i, end=" ")

2 5 8 11 14 17 

In [49]:
for i in range(-20, 20, 8):
    print(i, end=" ")

-20 -12 -4 4 12 

In [50]:
for i in range(100, -30, -25):
    print(i, end=" ")

100 75 50 25 0 -25 

## The `while` loop
- Conditional loop
    - The while loop iterates as long as the bool_exp is True
    - Once the bool_exp becomes False the loop ends
- The *while* loop can be used as *infinite loops*
- The *while* loop is called as event loop in GUI applications. As long as the loop runs the GUI application runs, once the event loop (while loop) ends the application closes.

###### Syntax for `while` loop
```
while bool_exp:
    block
```

In [51]:
# the infinite loop where bool_exp is always True
'''
while True:
    print("Hello, World!")
'''

'\nwhile True:\n    print("Hello, World!")\n'

In [52]:
# another infinite loop where bool_exp is always True
'''
while 10 > 5:
    print("Good evening!")
    print("~" * 13)
'''

'\nwhile 10 > 5:\n    print("Good evening!")\n    print("~" * 13)\n'

In [53]:
# The following loop never runs because the bool_exp is False from the start
while False:
    print("Good evening!")

In [54]:
# The following loop never runs because the bool_exp is False from the start
while 100 > 200:
    print("--------")

**NOTE**: The `while` loop does all that `for` loop does, and more.

In [55]:
# to prinit 1, 2, 3, 4, 5 using for loop
# start = 1; stop = 6 (5 + 1), step = 1
#	range(1, 6, 1)
#	range(1, 6)

for i in range(1, 6):
	print(i, end=" ")

1 2 3 4 5 

In [56]:
# to prinit 1, 2, 3, 4, 5 using for while
# start = 1; stop = 6 (5 + 1), step = 1

i = 1					# start
while i < 6:			# stop
	print(i, end=" ")
	i = i + 1			# step

1 2 3 4 5 

In [57]:
# print -4, -1, 2, 5, 8, 11 using for and while loops
# start = -4; stop = 12 (11 + 1); step = 3

for i in range(-4, 12, 3):
	print(i, end=" ")

i = -4                  # start
while i < 12:           # stop
	print(i, end=" ")
	i = i + 3           # step

-4 -1 2 5 8 11 -4 -1 2 5 8 11 

### The `break` statement
- We use *break* statement to break (end) a loop early, *for* or *while* loop
- Normally, we write *break* statement inside *if* statement to give the reason for breaking the loop

In [1]:
for n in [12, 85, 44, 35, 61, 90, 88, 11]:
    print(n, end=" ")
    if n % 7 == 0:
        break

12 85 44 35 

In [2]:
for n in [12, 85, 44, 35, 61, 90, 88, 11]:
    if n % 7 == 0:
        break
    print(n, end=" ")

12 85 44 

In [3]:
for i in range(1, 11):          # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    if i % 3 == 0:
        break
    print(i)

1
2


In [4]:
emp_timings = ["raju", "kiran", "naveen", "prasad", 
               "uday", "bhaskar", "sunil", "anil", "kavi",
               "mahesh", "chandu", "umesh", "dinesh"]
for name in emp_timings:
    print("Name:", name)
    if name == "uday":
        break

Name: raju
Name: kiran
Name: naveen
Name: prasad
Name: uday


#### The `continue` statement
- You can skip an iteration using the *continue* statement inside a loop
- When an iteration is skipped the control jumps to the beginning of the loop
- Always the *continue* statement is coded inside *if* statement, because you need a reason to skip an iteration

In [5]:
for x in range(1, 11):          # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    if x % 2 == 0:
        continue
    print(x, end=" ")

1 3 5 7 9 

In [6]:
for x in range(1, 11):          # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    if x % 3 == 0:
        continue
    print(x, end=" ")

1 2 4 5 7 8 10 

In [7]:
emp_timings = ["raju", "kiran", "naveen", "prasad", 
               "uday", "bhaskar", "sunil", "anil", "kavi",
               "mahesh", "chandu", "umesh", "dinesh"]
for name in emp_timings:
    if "i" not in name:
        continue
    print(name, end=" ")

kiran sunil anil kavi dinesh 